ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+      Assert.fail("Instantiation of GetReadableStreamChannel operation with an invalid blob id must fail");
+    } catch (RouterException e) {
+      Assert.assertEquals("Unexpected exception received on creating GetBlobOperation", RouterErrorCode.InvalidBlobId,
+          e.getErrorCode());
+    }
+
+    // test a good case
+    GetBlobOperation op =
+        new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            operationCallback, operationCompleteCallback, time);
+
+    Assert.assertEquals("Callbacks must match", operationCallback, op.getCallback());
+    Assert.assertEquals("Futures must match", operationFuture, op.getFuture());
+    Assert.assertEquals("Blob ids must match", blobIdStr, op.getBlobIdStr());
+  }
+
+  /**
+   * Put blobs that result in a single chunk; perform gets of the blob and ensure success.
+   */
+  @Test
+  public void testSimpleBlobGetSuccess()
+      throws Exception {
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    for (int i = 0; i < 10; i++) {
+      // blobSize in the range [1, maxChunkSize]
+      blobSize = random.nextInt(maxChunkSize) + 1;
+      doPut();
+      getAndAssertSuccess();
+    }
+  }
+
+  /**
+   * Put a blob with no data, perform get and ensure success.
+   */
+  @Test
+  public void testZeroSizedBlobGetSuccess()
+      throws Exception {
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    blobSize = 0;
+    doPut();
+    getAndAssertSuccess();
+  }
+
+  /**
+   * Put blobs that result in multiple chunks and at chunk boundaries; perform gets and ensure success.
+   */
+  @Test
+  public void testCompositeBlobChunkSizeMultipleGetSuccess()
+      throws Exception {
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    for (int i = 2; i < 10; i++) {
+      blobSize = maxChunkSize * i;
+      doPut();
+      getAndAssertSuccess();
+    }
+  }
+
+  /**
+   * Put blobs that result in multiple chunks with the last chunk less than max chunk size; perform gets and ensure
+   * success.
+   */
+  @Test
+  public void testCompositeBlobNotChunkSizeMultipleGetSuccess()
+      throws Exception {
+
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    for (int i = 0; i < 10; i++) {
+      blobSize = maxChunkSize * i + random.nextInt(maxChunkSize - 1) + 1;
+      doPut();
+      getAndAssertSuccess();
+    }
+  }
+
+  /**
+   * Test the case where all requests time out within the GetOperation.
+   * @throws Exception
+   */
+  @Test
+  public void testRouterRequestTimeoutAllFailure()
+      throws Exception {
+    doPut();
+    operationsCount.incrementAndGet();
+    GetBlobOperation op =
+        new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            null, operationCompleteCallback, time);
+    requestRegistrationCallback.requestListToFill = new ArrayList<>();
+    op.poll(requestRegistrationCallback);
+    while (!op.isOperationComplete()) {
```

Comment:
@nsivabalan On the contrary, this is exactly the way this should be written don't you think? Why would a `MockServer` say "timed out at the router"? That feels hacky.